dartagnan/src/main/java/com/dat3m/dartagnan/wmm/axiom/Irreflexive.java

```diff
@@ -24,26 +27,27 @@ public Irreflexive(Relation rel) {
     }
 
     @Override
-    public TupleSet getEncodeTupleSet(){
-        TupleSet set = new TupleSet();
-        rel.getMaxTupleSet().stream().filter(Tuple::isLoop).forEach(set::add);
-        return set;
+    protected Set<Tuple> getEncodeTupleSet(Context analysisContext) {
+        final RelationAnalysis ra = analysisContext.get(RelationAnalysis.class);
+        return ra.getKnowledge(rel).getMaySet().stream().filter(Tuple::isLoop).collect(toSet());
     }
 
     @Override
-    public BooleanFormula consistent(Set<Tuple> toBeEncoded, EncodingContext ctx) {
+    public BooleanFormula consistent(EncodingContext ctx) {
     	BooleanFormulaManager bmgr = ctx.getBooleanFormulaManager();
 		BooleanFormula enc = bmgr.makeTrue();
-        for (Tuple tuple : toBeEncoded) {
+        final RelationAnalysis ra = ctx.getAnalysisContext().get(RelationAnalysis.class);
+        final EncodingContext.EdgeEncoder edge = ctx.edge(rel);
+        for (Tuple tuple : ra.getKnowledge(rel).getMaySet()) {
```

Comment:
I think `toBeEncoded` used to be the encodeTupleSet. Here we iterate over the may set. Is this correct?

Reply:
The code does effectively iterate over the encode set (cause it looks for loops in the may set).
I think it would be nicer to iterate over `getEncodeTupleSet` to make the logic more clear, even if this creates a second set (which is only linear in the program, so the overhead is non-existent).

Reply:
Ok, although the coupled code was located 10 lines from here. dry